game/src/core/Game.java

```diff
@@ -475,10 +474,61 @@ public static Point randomTilePoint() {
    * @param elementTyp Type of the Tile.
    * @return Position of the Tile as Point.
    */
-  public static Point randomTilePoint(final LevelElement elementTyp) {
+  public static Optional<Point> randomTilePoint(final LevelElement elementTyp) {
     return currentLevel().randomTilePoint(elementTyp);
   }
 
+  /**
+   * Get all tiles from the current level.
+   *
+   * @return A Set containing all tiles in the current level.
+   */
+  public static Set<Tile> allTiles() {
+
+    return Arrays.stream(currentLevel().layout()) // Stream the layout (2D array)
+        .flatMap(Arrays::stream) // Flatten it into a stream of individual elements
+        .collect(Collectors.toSet()); // Collect the elements into a Set
+  }
+
+  /**
+   * Get all tiles of the specified type from the current level.
+   *
+   * @param elementTyp Type of the tiles to retrieve.
+   * @return A Set containing all tiles of the specified type in the current level.
+   */
+  public static Set<Tile> allTiles(final LevelElement elementTyp) {
+    Set<Tile> tiles = allTiles();
+    tiles.removeIf(tile -> tile.levelElement() != elementTyp);
+    return tiles;
+  }
+
+  /**
+   * Get all free tiles from the current level. A tile is considered free if no entity is present on
+   * it and it is accessible.
+   *
+   * @return A Set containing all free tiles in the current level.
+   */
+  public static Set<Tile> allFreeTiles() {
+    Set<Tile> tiles = allTiles();
+    // Remove the tile if an entity is on it or it is not accessible.
+    tiles.removeIf(t -> entityAtTile(t).count() > 0 || !t.isAccessible());
+    return tiles;
+  }
+
+  /**
+   * Get a random free tile from the current level. A free tile is a tile that is of type FLOOR and
+   * is not occupied by any entity and is accessible.
+   *
+   * @return An Optional containing a random free tile if available, otherwise an empty Optional.
+   */
+  public static Optional<Tile> freeTile() {
+    // create arraylist so i can shuffle the immutable list
+    List<Tile> floorTiles = new ArrayList<>(Game.allTiles(LevelElement.FLOOR).stream().toList());
+    // Shuffle the tiles for more randomness
+    Collections.shuffle(floorTiles);
+    return floorTiles.stream().findFirst();
```

Comment:
s.o.: teuer!

Vielleicht nicht wirklich richtig random, aber was wäre mit `findAny()`? Das soll ja *irgendein* Element finden ... Aber vermutlich liefert das auch einfach nur das erste gefundene Element zurück.

Ich würde aber unbedingt nochmal schauen, ob das nicht irgendwie beim Stream-Abschluss berechnet werden kann.

Reply:
Und Du hast Recht - das ist am Ende dennoch im Mittel viel teurer, als einfach nur eine rekursive Random-Funktion aufzurufen. Ich hatte tatsächlich eine Breitensuche im Kopf, die deutlich effizienter ist als der hier implementierte Gedanke: Gib mir alle passenden Tiles, und ich nehme davon dann eines.

Reply:
Beim drüber lesen fällt mir auf, ich brauch den ganzen Shuffle KRams eigentlich nicht. Ich kann mir mit der Listen länge einen Random Index ausspucken lassen, da war ich irgendwie zu komplex unterwegs.
Dann wäre die Komplexität "nur"  2D-Array in Liste "umrechnen" ( O(n)?) 

Breitensuche wäre im mittel effizienter, im worst Case auch O(n). Code Komplexität ist höher. Aber ich bau das mal ein, einfach zum gucken

Reply:
ja, du hast dann immer o(n). 

bfs ist exponentiell (mögliche verzweigungen hoch ebene der lösung - 1) - im worst case musst du da auch alle tiles ablaufen, aber im normalfall hast du deutlich weniger. 

beispielsweise bei 4 verzweigungen (möglichen nachbarfeldern) und in der dritten suchebene findest du was, dann bist du bei 4^(3-1), also 16 besuchten tiles (und drei ebenen sind bei uns recht viel). in deiner variante hast du **immer** 100 oder 200 oder was tiles. und diese methode wird häufig aufgerufen ...

Reply:
Ich hab mal ein BFS implementiert. schaus dir an.

Reply:
mach ich morgen ... irgendwie ist am ende des tages immer noch zu viel zu tun übrig ...

Reply:
@AMatutat Ich musste eine Weile nachdenken - in den mir bekannten BFS-Varianten wird ein Knoten erst dann als "bearbeitet" markiert, wenn ich ihn aus der Queue rausnehme und seine Nachfolger expandiere. Ich glaube aber, die hier vorliegende Variante könnte funktionieren, da wir keine Wege suchen und nur ein freies Feld.

Zur konkreten Umsetzung:

- Der Code selbst macht Low-Level Matrix-Kram auf dem Level-Layout. Das sollte in die Klasse LevelUtils verschoben werden, und hier in Game nur die Weiterleitung.
- Gibt es einen Grund, warum Du immer neue Tupel-Objekte erzeugst und in die Queue packst und nicht einfach die jeweilige Tile-Referenz? Das Abfragen der Koordinaten ist bei Verwendung der Tile-Referenz mind. ein Methodenaufruf (wie jetzt auch), nur in der aktuellen Variante kommt noch das Erzeugen des neuen Tupel-Objekts hinzu?
- Gibt es einen Grund, nicht einfach eine Doppelschleife `for (int i: {-1,1}) {for (int j: {-1,1} { machewas(layout[i][j]); }}` zu nutzen? Das wäre irgendwie leichter lesbar, und die diagonalen Felder würde man ja sowieso irgendwann untersuchen (auch wenn man so nicht laufen kann).
- Das Berechnen der Nachbarn sollte in LevelUtils (oder Tile) sein und ein Set oder was zurückliefern.

Reply:
> Gibt es einen Grund, nicht einfach eine Doppelschleife for (int i: {-1,1}) {for (int j: {-1,1} { machewas(layout[i][j]); }} zu nutzen? Das wäre irgendwie leichter lesbar, und die diagonalen Felder würde man ja sowieso irgendwann untersuchen (auch wenn man so nicht laufen kann).

Du hast recht, mach ich hier (warum auch immmer) falsch. Hab mich vermutlich verhaspelt. Funktionieren tut es auch so, aber ich fix das. 


>Gibt es einen Grund, warum Du immer neue Tupel-Objekte erzeugst und in die Queue packst und nicht einfach die jeweilige Tile-Referenz? 

Jetzt wo ich nochmal drauf gucke, wäre `Coordinate` der richtige Typ. `Tile#coordinate` gibt mir die Coordinate und ich kann genau diese Referenz speichern. Dann hab ich was ich brauche. Man könnte auch direkt auf der Tile-Referenz agieren, aber irgendwie ist mir das. zu Matrjoschka beim Code lesen.

 > Gibt es einen Grund, nicht einfach eine Doppelschleife for (int i: {-1,1}) {for (int j: {-1,1} { machewas(layout[i][j]); }} zu nutzen? Das wäre irgendwie leichter lesbar, und die diagonalen Felder würde man ja sowieso irgendwann untersuchen (auch wenn man so nicht laufen kann).

Fand meine Variante tatsächlich lesbarer. 
Wenn ich Coordinate anstelle von Tupel nutze, könnte ich sogar `next = currentPosition + deltaVektor[i]`machen.

Reply:
@cagix guck mal rüber

Reply:
Oh, Postings crossing. Hmmm, ich schau gleich mal auf den Code.

Reply:
@AMatutat Ich glaube, das könnte jetzt so passen. So richtig ganz sicher bin ich noch nicht, ob LevelUtils wirklich die richtige Plattform für die ganzen neuen Berechnungen ist, aber Du hättest sicher was gesagt, wenn Du was besseres gesehen hättest, oder?